### ISSUE TITLE
evm_circuit/RestoreContextGadget: index out of bounds

### ISSUE THREAD

**Issue by pinkiebell** [privacy-scaling-explorations/zkevm-circuits#598]
After running cargo test w/ prover enabled on latest zkevm-chain I encounter:
`thread 'tokio-runtime-worker' panicked at 'index out of bounds: the len is 3 but the index is 3', /zkevm-circuits/zkevm-circuits/src/evm_circuit/util/common_gadget.rs:196:21`

Here is a standard block trace for that block: https://bafybeidmdglhpbv7fkvslh7sh4mpjqtwl2k4lagkkogqaesuiwacgvoeri.ipfs.infura-ipfs.io/TRACE

**Comment by pinkiebell** [comment]
this a side effect of using STOP for unimplemented opcodes. Sorry for the noise!

**Comment by pinkiebell** [comment]
> > this is a side effect of using STOP for unimplemented opcodes.
> 
> Ah sorry I didn't realize that `STOP` is used for unimplemented opcodes somewhere else, I thought I already replaced all of that with `RETURN`, would you like to share which part causes this panic? Or it's just modification locally?

A local modification here: https://github.com/pinkiebell/zkevm-circuits/commit/d535c52c4d2a8f7289ac38c6db821c161777531f

**Comment by han0110** [comment]
Ah I see, replacing it with `RETURN` should fix this.

**Comment by pinkiebell** [comment]
Does it make sense to implement a DummyGadget for them instead of using RETURN?

**Comment by han0110** [comment]
Yeah it makes totally sense, then we don't need to update that again if `RETURN` is implemented someday.

`DummyGadget` is actually implemented [here](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/execution/dummy.rs), but seems not "dummy" enough for all unimplemented cases. @lispc Do you think it's okay to remove those stack operation checks and make `DummyGadget` for all unimplemented states usage? Or there is any thing depending on this check so it's necessary?

**Comment by pinkiebell** [comment]
I'm adding the missing opcodes at the moment :)

**Comment by pinkiebell** [comment]
I'm getting an error for `EXTCODESIZE`: `index out of bounds: the len is 0 but the index is 0`
where `extcodesize_gadget: DummyGadget<F, 1, 1, { ExecutionState::EXTCODESIZE }>,` .

Could this error originate from somewhere else higher up the callchain?

**Comment by han0110** [comment]
That's because all unimplemented opcodes in crate `bus-mapping` are using `dummy_gen_associated_ops` like:

https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/d59be68cb23ed5fbd7439f1a4d36535db21ed894/bus-mapping/src/evm/opcodes.rs#L83-L88

So no stack operations are generated :(

A naive approach to disable the stack operation check in `DummyGadget`, otherwise we need to implement the similar stack logic in `dummy_gen_associated_ops`.

**Comment by pinkiebell** [comment]
Thanks for the pointer! Was missing this one :)
